source/extensions/clusters/common/logical_host.cc

```diff
@@ -3,17 +3,72 @@
+    const envoy::config::endpoint::v3::LbEndpoint& lb_endpoint,
+    const Network::TransportSocketOptionsConstSharedPtr& override_transport_socket_options,
+    TimeSource& time_source)
+    : HostImplBase(lb_endpoint.load_balancing_weight().value(),
+                   lb_endpoint.endpoint().health_check_config(), lb_endpoint.health_status()),
+      LogicalHostDescription(
+          cluster, hostname, address,
+          // TODO(zyfjeff): Created through metadata shared pool
+          std::make_shared<const envoy::config::core::v3::Metadata>(lb_endpoint.metadata()),
+          locality_lb_endpoint.locality(), lb_endpoint.endpoint().health_check_config(),
+          locality_lb_endpoint.priority(), time_source),
+      override_transport_socket_options_(override_transport_socket_options) {
+  address_list_ = std::make_shared<AddressVector>(address_list);
+  ASSERT(address_list_->empty() || *address_list_->front() == *address_);
+  health_check_address_ =
+      resolveHealthCheckAddress(lb_endpoint.endpoint().health_check_config(), address);
+}
+
+void LogicalHost::setNewAddresses(const Network::Address::InstanceConstSharedPtr& address,
+                                  const AddressVector& address_list,
+                                  const envoy::config::endpoint::v3::LbEndpoint& lb_endpoint) {
+  const auto& health_check_config = lb_endpoint.endpoint().health_check_config();
+  // TODO(jmarantz): change setNewAddresses interface to specify the address_list as a shared_ptr.
+  auto health_check_address = resolveHealthCheckAddress(health_check_config, address);
+  auto shared_address_list = std::make_shared<AddressVector>(address_list);
+  absl::MutexLock lock(&address_lock_);
+  address_ = address;
+  address_list_ = shared_address_list;
+  ASSERT(address_list_->empty() || *address_list_->front() == *address_);
+  health_check_address_ = std::move(health_check_address);
+}
+
+HostDescription::SharedConstAddressVector LogicalHost::addressList() const {
+  absl::MutexLock lock(&address_lock_);
+  return address_list_;
+}
+
+Network::Address::InstanceConstSharedPtr LogicalHost::address() const {
+  absl::MutexLock lock(&address_lock_);
+  return address_;
+}
+
 Upstream::Host::CreateConnectionData LogicalHost::createConnection(
     Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
     Network::TransportSocketOptionsConstSharedPtr transport_socket_options) const {
-  const auto current_address = address();
-  const std::vector<Network::Address::InstanceConstSharedPtr>& address_list = addressList();
-
-  return HostImpl::createConnection(
-      dispatcher, cluster(), current_address, address_list, transportSocketFactory(), options,
+  Network::Address::InstanceConstSharedPtr address;
+  SharedConstAddressVector address_list;
+  {
+    absl::MutexLock lock(&address_lock_);
+    address = address_;
+    address_list = address_list_;
```

Comment:
Instead of direct access the data members here, maybe you can add a new method to copy both address_ and address_list and have the method protected by lock?  With that, you can also move the address_ and address_list_ from protected into private section.

Reply:
I could but it doesn't seem worth it; it's just 2 lines protected by a lock, and no other call-sites.

Reply:
You did have the version that returned a pair before, and that's fine but I decided that was more verbose than just taking the lock and copying out the data.